### PR TITLE
provide shardingSphere(4.0.0,4.0.1, 4.1.0 ,4.1.1) plugin.

### PR DESCRIPTION
**Pull request by yu199195**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

#4897

### CODE REVIEW

File: apm-sniffer/apm-sdk-plugin/pom.xml [hunk 1]
```diff
@@ -47,6 +47,7 @@
         <module>sharding-sphere-3.x-plugin</module>
         <module>sharding-sphere-4.x-plugin</module>
         <module>sharding-sphere-4.x-RC3-plugin</module>
+        <module>sharding-sphere-4.1.0-plugin</module>
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/pom.xml:50]: Could you merge these 4 as one module? The source codes are duplicated logically.

Comment (yu199195) [reply apm-sniffer/apm-sdk-plugin/pom.xml:50]: Each version of the package has a different path, so different versions need to be provided

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/pom.xml:50]: That is not my asking. I am talking on the maven module perspective.

Comment (yu199195) [reply apm-sniffer/apm-sdk-plugin/pom.xml:50]: Maybe they could merge into a module,like this:

sharding-sphere-plugin
       ---- sharding-sphere-3.x-plugin
       ---- sharding-sphere-4.x-plugin
       ---- sharding-sphere-4.x-RC3-plugin
       ---- sharding-sphere-4.1.0-plugin       

What do you think, maybe there will be more submodules in the future.

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/pom.xml:50]: Skywalking doesn't need the different plugin jars to identify the versions, your plugin instrumentation witness definition makes that. So, once you moved the codes into one plugin, you could use separate packages for different versions. Others wouldn't be loaded.

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/pom.xml:50]:
```
sharding-sphere-plugin
---- sharding-sphere-3.x-plugin
---- sharding-sphere-4.x-plugin
---- sharding-sphere-4.x-RC3-plugin
---- sharding-sphere-4.1.0-plugin
```

If you want, you could, but my point is, there are so many ShardingSphere plugin jars. Are they really so different?

Comment (yu199195) [reply apm-sniffer/apm-sdk-plugin/pom.xml:50]: sharding-sphere, In previous versions, the named paths of packages have changed, so have many jar, but after 4.0 ~5.0 , it to share one.   5.0 maybe provide annother jar.

Comment (tuohai666) [reply apm-sniffer/apm-sdk-plugin/pom.xml:50]: Hi Sheng,
Yes, the class name and method name are so different almost every ShardingSphere release.
As I know, 4.0.0 is different from 4.0.0-RC3, and 4.1.0 is differnent from 4.0.1.
For the 4.1.1 under releasing, same as 4.1.0, they can share a same plugin.

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/pom.xml:50]: I still hope you could move all codes in one jar. You could see in the current codebase, there is no plugin for other libs like this :)

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/pom.xml:50]: Anyway, this is purely my hope.

**Comment by wu-sheng** [comment]
Another question, we are closing to release the SkyWalking 8.0.0 in the next 2 days. Are you intending to add this in that version? @yu199195

File: apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/pom.xml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    <artifactId>apm-sharding-sphere-4.1.0-plugin</artifactId>
+
+    <packaging>jar</packaging>
+    <name>sharding-sphere-4.1.0-plugin</name>
+    <url>http://maven.apache.org</url>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <shardingsphere.version>[4.0.0, )</shardingsphere.version>
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/pom.xml:37]: We don't accept this. The version should be accurate. We were facing your version change broke this before.

File: apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/pom.xml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>mysql</groupId>
+            <artifactId>mysql-connector-java</artifactId>
+            <version>[2.0.14,6.0.6]</version>
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/pom.xml:44]: Same here.

File: apm-sniffer/apm-sdk-plugin/pom.xml (hunk 1 above)
Comment (yu199195) [reply apm-sniffer/apm-sdk-plugin/pom.xml:50]: Hi，Sheng,
I see your point, But ShardingSphere Changing the package path at release time is too frequent at 4.0.0-RC ,  But inevitably, there will be path changes during version updates, like springMVC3,4,5  They can't be combined into one.  It's just my idea...

File: apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/pom.xml (hunk 2 above)
Comment (yu199195) [reply apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/pom.xml:37]: ok,I will fix it.

**Comment by yu199195** [comment]
> Another question, we are closing to release the SkyWalking 8.0.0 in the next 2 days. Are you intending to add this in that version?

yes, If you can,  because shardingsphere 4.0.0 4.0.1 4.1.0 and 4.1.1 already released

**Comment by wu-sheng** [comment]
You will have the next 24 hours to make it ready.

**Comment by yu199195** [comment]
@wu-sheng  hi,sheng
CI is not pass , because :TracingContextTest.testSpanLimit:51 » NullPointer
CI on macOs is not pass,  because : org.apache.skywalking.oap.server.library.util.MultipleFilesChangeMonitorTest.test >> is fail

The above mistake should not be caused by my PR,What should I do?, hoping for your help.

thanks.

File: apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/pom.xml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+    <artifactId>apm-sharding-sphere-4.1.0-plugin</artifactId>
+
+    <packaging>jar</packaging>
+    <name>sharding-sphere-4.1.0-plugin</name>
+    <url>http://maven.apache.org</url>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <shardingsphere.version>[4.0.0,4.1.1]</shardingsphere.version>
```
[the remark this answers is not in the document]
Comment (yu199195) [reply apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/pom.xml:37]: ok, I init thought that versions of this range would work， if a single version is enough， i fix it.

[the remark this answers is not in the document]
Comment (yu199195) [reply apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/pom.xml:37]: ok, I init thought that versions of this range would work， if a single version is enough， i fix it.

File: apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/pom.xml [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            <artifactId>sharding-jdbc-core</artifactId>
+            <version>${shardingsphere.version}</version>
+            <scope>provided</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <artifactId>maven-deploy-plugin</artifactId>
```
[the remark this answers is not in the document]
Comment (yu199195) [reply apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/pom.xml:64]: i will fix it.

File: apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/src/main/java/org/apache/skywalking/apm/plugin/shardingsphere/v41/Constant.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * The type Constant.
+ */
+public final class Constant {
+    
+    /**
+     * The constant CONTEXT_SNAPSHOT.
+     */
+    public static final String CONTEXT_SNAPSHOT = "CONTEXT_SNAPSHOT";
```
[the remark this answers is not in the document]
Comment (yu199195) [reply apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/src/main/java/org/apache/skywalking/apm/plugin/shardingsphere/v41/Constant.java:29]: I think your suggestion is very good, in other PR, we will optimize it, this pr not work here,is also i think.

File: apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/src/main/java/org/apache/skywalking/apm/plugin/shardingsphere/v41/ExecuteInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+import org.apache.skywalking.apm.plugin.shardingsphere.v41.threadlocal.ContextThreadLocal;
+
+/**
+ * {@link ExecuteInterceptor} enhances.
+ * creating a local span that records the execution of sql.
+ */
+public class ExecuteInterceptor implements InstanceMethodsAroundInterceptor {
```
[the remark this answers is not in the document]
Comment (yu199195) [reply apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/src/main/java/org/apache/skywalking/apm/plugin/shardingsphere/v41/ExecuteInterceptor.java:35]: i will  fix it.

File: apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/src/main/java/org/apache/skywalking/apm/plugin/shardingsphere/v41/Constant.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/src/main/java/org/apache/skywalking/apm/plugin/shardingsphere/v41/Constant.java:29]: In JDBC case, there is usually a common module, then you have different versions.

[the remark this answers is not in the document]
Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/src/main/java/org/apache/skywalking/apm/plugin/shardingsphere/v41/Constant.java:29]: sharding-sphere-plugin
---- sharding-sphere-common
---- sharding-sphere-3.x-plugin
---- sharding-sphere-4.x-plugin
---- sharding-sphere-4.x-RC3-plugin
---- sharding-sphere-4.1.0-plugin

[the remark this answers is not in the document]
Comment (yu199195) [reply apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/src/main/java/org/apache/skywalking/apm/plugin/shardingsphere/v41/Constant.java:29]: sheng，we will provide other PR optimize it，include test case 。 this pr work it，maybe is too many。

[the remark this answers is not in the document]
Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/src/main/java/org/apache/skywalking/apm/plugin/shardingsphere/v41/Constant.java:29]: OK. Then please update the changes file in the root folder. Once this gets being merged, this is included in the 8.0.0 release.

[the remark this answers is not in the document]
Comment (yu199195) [reply apm-sniffer/apm-sdk-plugin/sharding-sphere-4.1.0-plugin/src/main/java/org/apache/skywalking/apm/plugin/shardingsphere/v41/Constant.java:29]: thanks，I have complied with your Suggestions and review . Now  the code is update，and summited.

**Comment by wu-sheng** [comment]
1. Supported-list.md
1. https://github.com/apache/skywalking/blob/master/CHANGES.md#java-agent

Please update both files.

**Comment by yu199195** [comment]
> 1.Supported-list.md
> https://github.com/apache/skywalking/blob/master/CHANGES.md#java-agent

1.suported-list.md alreay commit, 
2. changes is pr : https://github.com/apache/skywalking/pull/4900

**Pull request merged**